### PR TITLE
Added Parameter callbacks

### PR DESCRIPTION
**Pull request by MathewMacDougall**
### Description

Allows callback functions to be registered with Parameter objects. These functions will be called when the Parameter's value changes.

### Testing Done

unit tests

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom`
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/thunderbots/software/util/parameter/parameter.h [hunk 1]
```diff
@@ -233,6 +251,10 @@ class Parameter
     // Store the namespace of the parameter
     std::string namespace_;
 
+    std::mutex callback_mutex_;
```
Comment (garethellis0) [inline src/thunderbots/software/util/parameter/parameter.h]: could we group with with the `value_mutex_`?

**Review by garethellis0 — CHANGES_REQUESTED** [review]

File: src/thunderbots/software/util/parameter/parameter.h (hunk 1 above)
Comment (MathewMacDougall) [reply src/thunderbots/software/util/parameter/parameter.h]: fixed

**Review by akhilveeraghanta — APPROVED** [review]
LGTM  :tada:

**Review by garethellis0 — APPROVED** [review]
LGTM :+1: :tada:

**Pull request merged**